### ISSUE TITLE
Full RTL direction support

### ISSUE THREAD

**Issue by Nirlah** [facebook/lexical#2610]
The following is a short overview of bi-directional editing surfaces and my suggestion for improving Lexical’s implementation. _(Continued from the [Discord discussion](https://discord.com/channels/953974421008293909/955972012541628456/993985520747155476))_

### Overrview
Modern editing surfaces’ default behavior today is to support bi-directional content by setting the direction based on the first letter’s language. This can be seen in the current Lexical implementation, _Apple Notes_, _Gmail_, and _Google Docs_:

https://user-images.githubusercontent.com/1909713/177728347-054678d1-cf3e-47d5-817e-c5041ff144ef.mp4

One implementation detail crucial for longer-form and complex content (like a letter or an article) is maintaining the previous paragraph setting if the user manually sets it. _Microsoft Office Word_ gets this correctly:

https://user-images.githubusercontent.com/1909713/177722676-d603b861-daad-42b4-8378-f053bc455fa9.mp4

_The recordings demonstrate Hebrew, but the behavior is similar in Arabic._

### Implementation suggestions

The following are my suggestions for scalable bi-directional support in Lexical:
1. Maintain the current dynamic evaluation of direction as the default behavior.
2. Add support for manually toggling/setting the node’s direction.
3. If a node has a `direction` manually set, it should be respected. When parsing a serialized editor state, the current implementation dynamically evaluates the content, disregarding the provided configuration.
4. One (or both) of the following will allow composing more complex content (letter/article) without constant friction:
  4.1. If the root node’s direction is _manually_ set, it should be the default direction for child nodes.
  4.2. When creating the following paragraph, if the previous paragraph has a manual direction set, it should be copied into the next paragraph.

**Comment by Nirlah** [comment]
> > If a node has a direction manually set, it should be respected. (When parsing a serialized editor state, the current implementation dynamically evaluates the content, disregarding the provided configuration.)
> 
> QQ: Why it would not respect manual direction in case of deserializing, e.g. if you save editor state and then init new editor with serialized state wouldn't you want to keep that manually set direction?

Yes, the parenthesis describes the current Lexical implementation that ignores the node’s `direction` value when deserializing.

**Comment by milkrong** [comment]
what is status of this features?

**Comment by sopost-kirt** [comment]
Doesn't seem to have been fixed. Just adding my 2 cents here, it's currently impossible to write a sentence that starts in Arabic and continues in Latin, or vice versa. This is because the direction is set using [regex](https://github.com/facebook/lexical/blob/main/packages/lexical/src/LexicalConstants.ts#L95-L104) to [detect](https://github.com/facebook/lexical/blob/main/packages/lexical/src/LexicalUtils.ts#L186-L194) latin or Arabic, based on the first character in the string.

So it's not possible to have a RTL direction paragraph with the text:

```
حبيبتي hello
```

And conversely, it's not possible to have a LTR direction paragraph with the text:

```
hello حبيبتي
```

Unless I'm missing something, I'd consider this basic RTL support since it's quite typical to have English words in Arabic sentences or vice-versa, .e.g:

Coca-Cola مرحبا من
(Hello from Coca-Cola)

If we could at least disable the auto RTL detection, or better yet set it manually we could work with this. But fundamentally, I think RTL as a feature is incomplete without the ability to start a sentence in a direction opposite to the first character of it. This is pretty standard when editing rich text.

**Comment by james-atticus** [comment]
I've started taking a look at how the reconciler could be improved with respect to handling bidi text. What I would _like_ to do is push most of the responsibility for calculating direction to the browser (using `dir="auto"`), which already has a lot of logic built in. I have next to no experience with bidi text though, so would appreciate feedback on the below from you all.

The discussions so far, on this issue as well as https://github.com/facebook/lexical/issues/4930 and https://github.com/facebook/lexical/issues/7632, are mostly focused on paragraphs. The challenge comes from dealing with other block elements, like lists or tables. Ideally, the reconciler shouldn't have to know about child types of `ElementNode` (except for `RootNode`).

### Paragraphs

Summarising the discussions so far:
* If direction has been manually set on the paragraph, then that should take precedence, i.e. we set `dir="ltr/rtl"` on the paragraph DOM node.
* Else if no direction has been set on the root node, then the paragraph should default to `dir="auto"`. This will leave the browser to decide whether each block should be LTR or RTL individually based on the user agent and the text contents.
* Else direction _has_ been manually set on the root node, and that should apply to all paragraphs. I.e. the root `div` will have `dir="ltr/rtl"` but the paragraphs will not have `dir` set at all.

If a paragraph has its direction overridden and you press Enter to start a new paragraph, that would be copied over. You'd get similar behaviour if the direction is overridden at the root (new paragraph would inherit the direction). The main difference is that if direction has _not_ been overridden anywhere, then the cursor in a new paragraph will default to the left or right depending on the browser's setting. Once you start typing though, it will detect the directionality of the text and update accordingly.

![Image](https://github.com/user-attachments/assets/22cdc5fc-272d-4711-8a62-ee0a2f745312)

### Lists

This logic maps reasonably well to lists, assuming you want the whole list to be RTL or LTR (all the markers for a list to be aligned to the same side). If we apply the same logic as paragraphs to the list itself (i.e. `ul`/`ol`):
* If direction has been manually set on the list, that will apply to all list items.
* Else if no direction has been set on the root node, then the list would default to `dir="auto"`, meaning the whole list will be set to LTR or RTL depending on the text contents of the first list item.
* Else direction has been set on the root node, and that will apply to the list and its items too.

![Image](https://github.com/user-attachments/assets/0fdeb398-efd2-4e35-a67c-5e2be9c78f69)

**This is a break from the current behaviour in Lexical**, where each list item's direction is determined by its contents. Is this a reasonable change? If you have a mix of LTR and RTL in a list, would you expect it to switch for each list item or be consistent for the whole list? Worth noting that if desired, you could still set the direction on individual items.

For context, here's that same list in the Lexical playground:

![Image](https://github.com/user-attachments/assets/68a5c852-1db8-451c-b0e3-8e6804dcad6b)

### Tables

This is where I'm really unsure what the desired behaviour is. If we use the same logic as paragraphs above, then when the table is in `dir="auto"`, the whole table's direction is determined by the text in the first cell. That is, if the first cell contains RTL text, then the columns of the table will be RTL and the text in every cell will be RTL (unless overridden):

![Image](https://github.com/user-attachments/assets/d821175f-7098-4fe6-b12a-b2836480eb04)

If this is what you'd expect, then great, the implementation is pretty straightforward.

It gets trickier if we want the individual cells to each behave like the top-level root element (eg: within each cell, set paragraphs' direction based on their content). Below, I've set `dir="auto"` on the RTL text in the first paragraph (with `dir="auto" still set on the table). For whatever reason, this makes the browser treat the table as LTR, presumably because that's the default for my user agent.

![Image](https://github.com/user-attachments/assets/a60337ab-cc61-4fb3-80d2-fa3b6876164c)

For those with more bidi experience, **what's the preferred behaviour with mixed-direction tables?**:
* set the whole table to LTR/RTL based on the contents of the first cell; all content has the same direction
* the table's direction is based on the browser default (could be RTL or LTR); content within cells has direction calculated individually

Worth noting that with the second approach, it might work to have a mutation listener set the overall direction of the table based on the calculated direction of its contents.

**Comment by james-atticus** [comment]
@EmilianoMorghen @Nirlah Tagging as you reacted to the post - do you have any concerns, questions, or suggestions for the approach outlined above, or are you happy with it as it is?

As I said, I don't have real-world experience with bidi text, so what I've outlined is a bit of a guess. Want to make sure it's fit-for-purpose before starting implementation.

**Comment by Pourianof** [comment]
@james-atticus Thanks for your attention for fixing this problem. I'm not familiar with bidi algorithm so much, so I try to say my opinion as a tester or user.
About list, as my experiences, its enough that direction get apply to whole list mean that all items get same direction, it help the readability.
But why not give the option to user to have freedom to choose how set the direction for each part? does it make implementation hard? For example, when I worked with the playground in my project, I encounter a problem with RTL text contains 2 LTR word which separated with the comma or colon or other punctuations which caused awkward direction:

![Image](https://github.com/user-attachments/assets/6979e831-1222-4e5c-b0e1-8f203fed7bf0)
which expected to be like:

![Image](https://github.com/user-attachments/assets/634bc851-1aa4-4d31-87ad-9c2958de72f4)

some solution maybe is that user handle it manually and wrap the selection part in a node with desired direction or detect punctuations that not in English set then consider it as the letter in another language and make it respect to current direction

About tables the programs like Microsoft Word act this way which table cells inherit direction from last block but each cells could directed individually. But I think set direction automatically for each cell via inputs make it more friendly to avoid user exhaustion from directing cells or maybe add some options beside RTL and LTR for whole table direction

At the end, I apologize for my poor English and laziness for not using chatgpt

**Comment by james-atticus** [comment]
Thanks @Pourianof - this is much more about building an algorithm that suits user expectations, not the other way around, so that's all useful context.

### Bidi within a paragraph

For setting direction for each part in a paragraph, the browser doesn't have a good way to do what you're asking short of splitting the text up manually. For example, this HTML:

```
<ul dir="rtl">
  <li value="1" dir="rtl">
    <span data-lexical-text="true" dir="rtl">مرحبا Developer: Exception</span>
  </li>
</ul>
```

Produces this output:

![Image](https://github.com/user-attachments/assets/16c028bc-c738-4b20-b69b-39ade5dea2fa)

But if I split the spans and manually set `rtl` on each, then you get something closer to what you're asking for, with the colon on the left of "Developer":

```
<ul dir="rtl">
  <li value="1" dir="rtl">
    <span data-lexical-text="true" dir="rtl">مرحبا Developer: Exception</span>
    <strong data-lexical-text="true" dir="rtl">Developer:</strong>
    <span data-lexical-text="true" dir="rtl"> Exception</span>
  </li>
</ul>
```

![Image](https://github.com/user-attachments/assets/3c20c72b-f9cf-4bb7-adef-5b8a76deff15)

That said, I think I'll have to leave this as out-of-scope for the current proposal. Lexical only supports setting direction on `ElementNode`, and extending that to `TextNode` feels like a separate, much bigger project. If someone wanted to pick that up in future, [this article](https://learn.microsoft.com/en-us/dynamics365/fin-ops-core/dev-itpro/user-interface/bidirectional-support) from Microsoft about how Word does bidi text could be useful.

### Tables

To confirm, you would prefer the behaviour from the second screenshot above? I.e. this one:

![Image](https://github.com/user-attachments/assets/67da10fc-e42a-45dc-a4e2-0312ca45f1d4)

For the implementation, rather than setting `dir="auto"` for children of `RootNode`, maybe we should set it for any node whose parent is a root or shadow root, but is not a shadow root itself. `TableNode`, `TableRowNode` and `TableCellNode` are all shadow roots, so this would mean we end up with `dir="auto"` on the contents of the cell, but not the cell/row/table themselves. This would extend nicely to playground nodes like `LayoutContainerNode` and `CollapsibleContainerNode`.

As mentioned above, what this doesn't do is set the whole table to RTL automatically, however with a plugin you could watch for changes to the table's contents and set its direction manually. As an example, this table has `dir="auto"` for the paragraphs in the cells, and manually-set `dir="rtl"` for the table (and its scrollable wrapper):

![Image](https://github.com/user-attachments/assets/3860896e-6d9e-4811-a5da-6994bfb9397c)

**Comment by mhosken** [comment]
There is some good thinking here. I'm not a bidi specialist but it bites me often enough for me to be sensitised! I agree with much of the conclusions. I would say that I like the idea of keeping lists in the same order. I would also extend that to table columns. Columns typically provide structure to a table and so one would want the paragraph order to be the same within a column although it might differ across columns. The rows provide the data. As to column order, I think that follows the document direction.

For things like scripture. A document will be primarily in one language and script and so has an inherent document level bidi direction. I would suggest that the same is true for a dictionary. What language is the dictionary _in_ rather than about? That might help with a default direction. Of course direction can always be overridden either using markup or using bidi control characters in Unicode (which is how I would handle the Developer Exception: issue)

In summary I would suggest: keep it simple, consistent and obvious. The more special cases you have the more the user has to learn. Since this is a data entry tool, it doesn't have to be perfect. user and developer can learn together.

**Comment by sharoncorrell** [comment]
I agree with @sopost-kirt that there should always be ways to override the default direction. The first-character rule is a good default but not adequate without a way to override.

Personally I think that lists that are oriented with some elements right-aligned and others left-aligned looks strange, so that shouldn't be the default, but I supposed it should be made possible using overrides.

I think that according to the bidi algorithm you would expect all LTR text within a RTL paragraph to be fully LTR, because punctuation like colons takes on the directionality of the surrounding text. So I'd expect to see

<img width="720" height="69" alt="Image" src="https://github.com/user-attachments/assets/5cdeb337-e3e3-4a1d-92c9-05143f207168" />

and you'd need to do some fancy footwork to change it.

**Comment by james-atticus** [comment]
Thanks for the feedback, everyone. First pass of the proposed implementation is up at #7727.

**Comment by james-atticus** [comment]
@Pourianof @mhosken @sharoncorrell would you mind taking a look at the linked PR and testing out the editor, see if it works the way you would expect?

**Comment by GermanJablo** [comment]
I see the PR was merged. Does that fix all the behaviors noted in this issue? Or is there something that still needs to be addressed?

**Comment by etrepum** [comment]
I believe that the merged PR mostly only fixed the part where it was not possible to manually specify the direction of a node, and then delegated the default decision to the browser instead of regexes for explicit ltr/rtl everywhere.

To fix the more general payload issue with empty elements there needs to be some more infrastructure, maybe the editor should have some setting for dir on the root element (or assume it is set) and allow that to propagate down natively instead of explicitly specifying it everywhere that the direct parent isn't explicitly set to a value. Probably also makes sense to allow 'auto' to be an explicit value for `getDirection`/`setDirection`.

**Comment by smrifat1411** [comment]
@etrepum `dir="auto"` is dropped on import today — `$setDirectionFromDOM` only forwards `'ltr'`/`'rtl'`, so an element with `dir="auto"` lands as if it had no `dir` at all. The reconciler already speaks it (`$getReconciledDirection` returns `'ltr' | 'rtl' | 'auto' | null`); it's `ElementNode.getDirection`/`setDirection` and the serialized `direction` field that stop at `'ltr' | 'rtl' | null`.

Would you take a small PR that just adds `'auto'` as an explicit value there, scoped separately from the wider infrastructure this issue asks for? And since it widens the serialized `direction` type, is that better folded into #8602 than landed alongside it?

**Comment by etrepum** [comment]
Yes, open a PR for auto support.

**Comment by etrepum** [comment]
#8602 isn't a good place to change the serialization domain, would not recommend trying to fold it in. Allowing for an explicit 'auto' is a larger change to how bidi behaves, so while the code change is likely small (before or after #8602) it has a lot of potential side-effects that should be tested in isolation.

Because the pre-#8602 JSON serialization is not validating it is probably a larger and longer-term change than you'd think. It's not really possible to do in a backwards compatible way (without adding a new property or using NodeState for auto) since the `ElementNode.updateFromJSON` implementation does not validate the `dir` property at all, it will happily set out of bounds values. Collab has similar problems.
